components/dada-validate/src/validate/validator.rs

```diff
@@ -520,12 +543,26 @@ impl<'me> Validator<'me> {
                 self.add(validated::ExprData::Seq(validated_exprs), expr)
             }
             syntax::ExprData::Return(with_value) => {
-                if let Some(return_expr) = with_value {
-                    let validated_return_expr = self.validate_expr(*return_expr);
-                    return self.add(validated::ExprData::Return(validated_return_expr), expr);
+                match (self.code.return_type, with_value) {
```

Comment:
This is clever :) I was imagining we'd enforce this more dynamically, but I like the idea of checking it statically.